server/src/main/java/org/opensearch/ingest/IngestService.java

```diff
@@ -802,69 +822,87 @@ private void runBulkRequestInBatch(
         BiConsumer<Thread, Exception> onCompletion,
         IntConsumer onDropped
     ) {
-
         final Thread originalThread = Thread.currentThread();
-        final AtomicInteger counter = new AtomicInteger(numberOfActionRequests);
-        int i = 0;
+
+        int slot = 0;
         List<IndexRequestWrapper> indexRequestWrappers = new ArrayList<>();
+
+        // Walk through slot by slot. For each slot, go through inner slots (child index requests) and determine
+        // which pipelines to execute for each index request based on previously resolved pipelines.
+        // Inner/outer slots are used to map index requests to their position in the original bulk request list.
         for (DocWriteRequest<?> actionRequest : actionRequests) {
-            IndexRequest indexRequest = TransportBulkAction.getIndexWriteRequest(actionRequest);
-            if (indexRequest == null) {
-                if (counter.decrementAndGet() == 0) {
-                    onCompletion.accept(originalThread, null);
+            List<IndexRequest> childIndexRequests = new ArrayList<>();
```

Comment:
We can avoid ArrayList allocations even if the request is nit UpdateRequest
```suggestion
            List<IndexRequest> childIndexRequests = Collections.emptyList();
```